Fix __contains__ for atoms/residues in molecules. [closes #333]

lohedges opened the pull request:
This PR addresses #333, where atoms were not correctly being identified as belonging to the molecule from which they were extracted.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_molecule.py:
```diff
@@ -171,7 +171,7 @@ def __contains__(self, other):
 
         # Return whether the object comes from this molecule.
-        return self._sire_object.molecule() == other._sire_object.molecule()
+        return self._sire_object.molecule().number() == other._sire_object.molecule().number()
```
Is it worth overriding `__eq__` of the wrapped objects?

xiki-tempula replied:
@msuruzhon Would that be a Sire issue instead of a BSS issue?

msuruzhon replied:
Sorry I meant that `__eq__` of the `_SireWrappers.Molecule` should wrap this equals statement. Or is that not desirable in general?

xiki-tempula replied:
@msuruzhon I guess the core issue is that 

`BSS.molecule._sire_object == BSS.molecule._sire_object`

With regard to whether it is desirable, I just only @lohedges could comment.

lohedges replied:
In this case `other` could be an `Atom`, `Molecule`, or `Residue`, so this isn't a comparison at the molecule level. Calling `.molecule()` on a Sire object gets back the molecule to which the object belongs, but the comparison between those molecules doesn't appear to always be consistent. (For reference, you can already compare BioSimSpace molecules.)

```python
protein = BSS.IO.readMolecules(['/test/input/molecules/1jr5.crd', '../test/input/molecules/1jr5.top']).getMolecule(0)
ligand = BSS.IO.readMolecules(BSS.IO.glob("test/input/ligands/ligand01*")).getMolecule(0)

system = (ligand + protein).toSystem()

protein._sire_object.molecule() == system[1]._sire_object.molecule()
False
protein._sire_object.molecule().number() == system[1]._sire_object.molecule().number()
True

protein._sire_object
Molecule( 633 version 21 : nAtoms() = 1495, nResidues() = 90 )
system[1]._sire_object
Molecule( 633 version 1606 : nAtoms() = 1495, nResidues() = 90 )
```

I assume that the comparison is failing since the version numbers differ.

xiki-tempula replied:
@lohedges Make sense

lohedges replied:
I'm just trying to work out at which point the version number changes and why, and if it's important.

chryswoods replied:
Yes - that's correct. Sire tests equality by comparing the ID of the object and the version number. This enables us to do quick tests for equality without having to run through and compare all of the data in the molecule.

A change in version number implies that `.edit().set(...).commit()` has been called on the molecule somewhere. This could be either because the molecule has been translated, or it has its properties changed etc.

In general, there are there levels of comparison you can do:

1. Just compare the ID numbers of the objects (e.g. molecule.number() == other.number()). This only guarantees that the molecules were the same at one point in the past (the molecule number does not change).
2. Compare the UID of the molecule.data().info() object (the MoleculeInfo). This is the layout of the molecule, i.e. names of atoms, arrangement into residues etc. Every layout that is the same has the same UID. Having the same molecule number and layout UID means that the molecule is the same and it has the same atoms, residues, etc
3. Compare the version number of the object. Every time the object changes the version number will increase. Two objects with the same ID number and version number are guaranteed to be exactly the same.

This is why a general ".contains" is difficult to implement. It is ambiguous, as do you mean "this molecule contains an atom with this name", or "this version of this molecule contains the same version of the atom".

lohedges replied:
Yes, I agree with what @chryswoods says. Perhaps it's not worth including this feature yet because of the ambiguity, or at least document the operator so the use case is clear.

xiki-tempula replied:
I think one way to go would be
`is` will check both version number and MolNum
'__eq__' will only check MolNum?

lohedges replied:
Currently it's the opposite, which makes more sense to me, despite still being ambiguous. As mentioned above, this is because, when creating a system, atoms and residues are renumbered to be unique and in ascending order. This results in the following situation:

```python
s = BSS.IO.readMolecules("amber/ala/*")
ss = (s[1] + s[0]).toSystem()

# Check by number only.
s[0] in ss
True

# Check by number and version.
s[0] == ss[1]
False
```

It might be possible to move the renumbering code into `Process.Somd`, so that this is done once on setting things up for a SOMD simulation. I'll need to check that this doesn't cause other issues, though.

chryswoods replied:
`__eq__` implies equality. You could, in theory, change both the layout and the properties of a molecule while retaining the same molecule number. This would, in the worst case, mean you could have `__eq__` give true for two molecules that have different numbers of atoms. I would prefer something like

```
atom.molecule().number() in system
```

(i.e. you are explicitly asking if this molecule number is in a system)

I remember that Lester mentioned that he had a general "ID", so it could be

```
atom.ID() in system
```

xiki-tempula replied:
@lohedges Sorry, I mean `is` not `in`. The `is` could have a different behaviour to '__eq__'

chryswoods replied:
I see what you are saying, and it would make sense if this `__eq__` and `is` was used consistently throughout the rest of the code. However, because C++ doesn't have an `is`, it would be really hard to keep this use consistent. It would also be a potential source of surprise for people who didn't know about `is`. In general, I prefer `__eq__` to mean "these two things are identical - they are equal to each other". You can have two objects held in different bits of memory that have the same data, so would be `__eq__` but would not be `is` (`is` means that the two variables point to the same piece of memory).

I know it is a little annoying. But these different levels of comparison exist because there are different types of equality questions that people could ask, i.e. "these are exactly the same molecules", "these molecules have exactly the same data", "these two molecules have the same internal layout" etc.

Following "explicit is better than implicit", I prefer using `object in container` to mean the exact same data of object is in the container. Then `object.ID() in container` to mean that an object with the same ID is in the container. And then, `object.layout() in container` to mean that the container has an object with the same internal layout. (note that all three of these are pseudocode - we don't have a `object.layout()` function, although it may be a good thing to add)

lohedges replied:
I think I'll just try moving this bit...

```python
self._sire_object = _SireIO.renumberConstituents(self._sire_object, num_mols)
```

into `setup` method of `Process.Somd`. This way both `in` and `eq` will behave in _exactly_ the same way, i.e. checking for a molecule with the same number and version.

I'll just need to make sure that there aren't any unintended consequences of doing this, i.e. any other things that assume unique numbering elsewhere.

lohedges replied:
This is actually more complicated that I thought, and will require and update to the C++ wrapper that does the renumbering.

lohedges replied:
I've added some updates to Sire [here](https://github.com/michellab/Sire/tree/fix_bss_333) that resolve the issue with moving the renumbering functionality into the SOMD setup. (Previously the coordinate/velocity update wrapper assumed that the system had been renumbered, where now we leave the original system numbering untouched.) This seems to work, and allows us the `__contains__` operator to work on molecules, not molecule numbers.

I'll still have a think about how we want to deal with ambiguity, i.e. by providing the option have different levels of comparison, as suggested by @chryswoods. (Check only by number, or by number and version, etc.)

lohedges commented:
Okay, as mentioned in the thread above I've now removed the requirement that atom and residue numbers are unique and in ascending order, moving this entirely into the SOMD process setup. (Applying this to a copy of the original system.) This means that `__contains__` now works as expected when molecules haven't been edited. For now I think that this is the correct behaviour, i.e. `==` or `in` only return `True` for the _exact_ same molecule. If you want to check whether a molecule with the same _number_ is in a system, then you can use the search functionality, i.e. `system.search("molnum 42")`, then check the results.

Apologies for bundling them here, but I've also added a few simple additions for @xiki-tempula's work, along with appropriate unit tests.

* Apply units to items in a container, i.e. you can now do things like:
```python
[1, 2, 3] * BSS.Units.Length.angstrom
[1.0000 A, 2.0000 A, 3.0000 A]
```
* Allow the extraction of residues/atoms from a system by their _absolute_ index, i.e. you don't need to extract a molecule then get the residue/atom by its relative index.
```python
# Return the 20th residue.
residue = system.getResidue(20)

# Return the 100th atom.
atom = system.getAtom(100)
```

Please note that _all_ macOS CI builds will currently fail due to [this](https://github.com/michellab/Sire/issues/386) issue. Hopefully we will find a fix soon.

lohedges commented:
Actually, the macOS builds work, since there's nothing to check `BioSimSpace.Process.Somd.getSystem()`. It will fail when run, however, since the Sire version won't contain the appropriate wrapper.

xiki-tempula commented:
@lohedges Hi Lester, I wonder if this PR is ready for review? Miro is absent this week so I will be reviewing the PR.

lohedges commented:
Hi there, I'm away today. The issue isn't the BioSimSpace code, rather that it requires an update to Sire which isn't yet available on macOS due to a build issue that still needs to be fixed. I should be able to look at this tomorrow.

If you just need the atom extraction code you could always look at what I've done and merge/copy to your feature branch so that you can access that functionality (which doesn't need Sire to be updated) ahead of time.

xiki-tempula commented:
@lohedges Thanks. Can I assume that the BSS part is complete? So I don't need to resolve conflict later on.

lohedges commented:
I assume it will be okay. If not, I doubt there will be much to amend. I can always help you if there are problems.

xiki-tempula commented:
@lohedges Excellent, thanks.

lohedges commented:
@msuruzhon: We have now fixed the macOS build issue for Sire and have 2022.3.0 release and conda packages. Let me know when you're happy with this and I'll merge in to devel. This should resolve a few outstanding issues that are required for @xiki-tempula's work.

Cheers.

xiki-tempula commented on doc/source/compatibility.rst:
```diff
@@ -4,20 +4,30 @@
 `GROMACS <http://www.gromacs.org>`__
 ====================================
 
+* 2021 series
```
Do we wish to support 2022 as well?

lohedges replied:
Yes, as stated in the documentation it's just that we've not _fully_ tested against 2022 yet. There is a section stating that this does not mean that it won't work against more recent versions, only that it hasn't been fully battle hardened. We generally update at the end of the year for GROMACS, i.e. when we haven't experienced issues for _all_ releases from that year.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_system.py:
```diff
@@ -1702,6 +1758,47 @@ def _set_water_topology(self, format, property_map={}):
                 self._sire_object = _SireIO.setGromacsWater(
                         self._sire_object, water_model, property_map)
 
+    def _set_atom_index_tally(self):
```
Are these always regenerated on system modification?

lohedges replied:
It's called internally by whatever functions need the tally information, i.e. it's always called as the first thing within `getAtom`, or when getting the index of an atom with `getIndex`. It will be inconsistent if the system is updated, but is always consistent at the point of use.

msuruzhon commented on test/Types/test_types.py:
```diff
@@ -105,3 +105,16 @@ def test_from_sire_unit(Type):
+    [1, 2, 3] * my_type
+
+    # Apply the type to a tuple.
+    (1, 2, 3) * my_type
```
Should we assert that the result is actually of the desired type?

lohedges replied:
Yes, that's a good idea. I've now added the following:

```python
    # Apply the type to a list.
    container = [1, 2, 3] * my_type

    # Assert we have a container of the original type.
    assert all(isinstance(x, Type) for x in container)

    # Apply the type to a tuple.
    container = (1, 2, 3) * my_type

    # Assert we have a container of the original type.
    assert all(isinstance(x, Type) for x in container)
```

msuruzhon commented on test/_SireWrappers/test_system.py:
```diff
@@ -144,3 +144,17 @@ def test_contains_combine(system):
     assert m1 in s0
     assert m1 in s1
+
+def test_get_atom(system):
```
Can we add unit tests for calling `getAtom()`, modifying the system, and calling `getAtom()` again, just to make sure it works as expected?

lohedges replied:
I've updated the test so that we delete some molecules from the system, then make sure that `getAtom()` still works, i.e. it maps to the new molecule positions, not the old ones.

msuruzhon commented on python/BioSimSpace/Sandpit/Exscientia/Types/_area.py:
```diff
@@ -118,6 +118,14 @@ def __init__(self, *args):
+            container = []
+            for item in other:
+                container.append(self.__mul__(item))
+            return container
```
Should the container be of the same type as `other`?

lohedges replied:
Yes, that's a good point. I'll make sure it either returns a list or tuple depending on the original container.

msuruzhon commented on python/BioSimSpace/_SireWrappers/_system.py:
```diff
@@ -1702,6 +1758,47 @@ def _set_water_topology(self, format, property_map={}):
+            num_atoms = 0
+            for num in self._sire_object.molNums():
+                self._atom_index_tally[num] = num_atoms
+                num_atoms += self._sire_object.molecule(num).nAtoms()
```
Should these be checked against the `Sire` `molNums` or the `BioSimSpace` numbers which follow the order of the molecules? I imagine they are supposed to always be the same but in case they are not, this will result in unexpected behaviour?

lohedges replied:
It should always be the same but I'll check.

lohedges replied:
Yes, this holds in my checks, i.e. performing various system modifications and making sure the `MolNums` are reconstructed in the correct order, depending on how the System was created. This did raise one small bug, where adding a `Molecules` object to a `Molecule` resulted in the molecule being added to the _end_ of the existing `Molecules` container, rather than a new container being returned with the `Molecule` at the start. This didn't affect any indexing operations on the resulting container, only that the items in the container were in an unexpected order to start with.

msuruzhon reviewed (changes requested):
Thanks @lohedges I have left a few comments and questions.

lohedges commented on python/BioSimSpace/Parameters/Protocol/_amber.py:
```diff
@@ -623,15 +623,6 @@ def run(self, molecule, work_dir=None, queue=None):
                                                      "make sure that your initial molecule has a "
                                                      "complete topology.")
 
-                    # If the original molecule was comprised of multiple chains, then we need
```
Because the conditional below has now been removed since it is redundant. The comment only applied to this section.

lohedges replied:
The comment only applied to the conditional block below, which has now been removed since it is redundant.

lohedges commented on python/BioSimSpace/Process/_somd.py:
```diff
@@ -212,6 +212,13 @@ def _setup(self):
+        # are unique, i.e. the residue number of the perturbed molecule.
+        # We store the renumbered system to use as a template when mapping atoms
+        # the those from the extracted trajectory frames in, e.g. getSystem().
+        self._renumbered_system = _SireIO.renumberConstituents(system._sire_object)
```
No, the behaviour should be identical. All I've done is removed the requirement that atom and residue indices are unique when modifying a System object in any way, e.g. adding molecules. This is slow when combining large systems. Instead, I've moved this into the SOMD process, where I store the renumbered system as an internal attribute. This is then used as a "reference" system for mapping back molecules when updating the coordinates and velocities. (Effectively all I've done is move the point at which we enforce the unique indexing constraint.)

lohedges commented:
Thanks for the comments, they were really useful. I think I've now responded to everything. Let me know if you have any further questions.

Cheers.

msuruzhon reviewed (approved):
Many thanks for that @lohedges, I am happy for this to be merged.
